### PR TITLE
Poll loop fixes

### PR DESCRIPTION
**Pull request by vext01**
### What does this PR try to resolve?

A couple of minor issues. See individual commits.

### How should we test and review this PR?

The existing test suite should have sufficient coverage.

### Additional information

(I have 5 apparently unrelated failures locally when running `cargo test`)

### CODE REVIEW

**Comment by Eh2406** [comment]
What does this do/accomplish? How would we know if we break this functionality in the future?

**Comment by vext01** [comment]
If your `fcntl()` call fails (and I have seen this happen) then you would be doing blocking IO in the poll loop, likely ending in a deadlock.

**Comment by vext01** [comment]
Also, getting the current flags and ORing in the non-block flag avoids destructively turning off any other flags the fd may already have set.

File: crates/cargo-util/src/read2.rs [hunk 1]
```diff
@@ -2,21 +2,28 @@ pub use self::imp::read2;
...
 mod imp {
+    use libc::{c_int, fcntl, F_GETFL, F_SETFL, O_NONBLOCK};
     use std::io;
     use std::io::prelude::*;
     use std::mem;
     use std::os::unix::prelude::*;
     use std::process::{ChildStderr, ChildStdout};
 
+    fn set_nonblock(fd: c_int) -> io::Result<()> {
+        let flags = unsafe { fcntl(fd, F_GETFL) };
```
Comment (ehuss) [inline crates/cargo-util/src/read2.rs:13]: There is some uncertainty whether or not F_GETFL takes an argument. https://stackoverflow.com/questions/25061656/when-is-the-arg-for-f-getfl-fcntl-command-required has an interesting discussion of the issue. The [Open Group](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fcntl.html) specification notes that the arg value is reserved for future growth, without specifying what that means or what it should be set to or whether or not it should be specified. [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=fcntl&sektion=2) specifically says it is ignored. Very strange.

**Comment by vext01** [comment]
> Can you say what system you are running on and how you have been seeing it fail in cargo?

To be clear. I've seen `fcntl` fail in other Rust programs (due to programming errors), not in Cargo.

But regardless, it's good practice to check for errors defensively so that if there is a problem, it happens as close to the cause as possible.

**Comment by vext01** [comment]
We good to go?

**Comment by ehuss** [comment]
Thanks! I'll go ahead and merge.  I'm a little uncomfortable with this change, but other than the uncertainty about F_GETFL's behavior, I don't know of any specific issues. I also generally avoid changing things if they aren't causing any issues.

@bors r+

**Pull request merged**